### PR TITLE
Factor out language_infot from languaget

### PR DESCRIPTION
**Pull request by peterschrammel**
Prerequisite for https://github.com/diffblue/cbmc/pull/2127 and https://github.com/diffblue/cbmc/pull/2098 to make access to language-specific methods more efficient.

Previously, using language-specific methods such as
from_expr() required creating a new languaget instance
which could be potentially costly.
This PR introduces a language_infot class that offers
language-specific methods that don't require parsing
functionality. language_infot is now used for language
registration (instead of language_entryt). The immutable
registered *_language_infot instances are shared among the
corresponding *_languaget instances that are requested by
the get_language_from_* functions.

### CODE REVIEW

File: src/clobber/CMakeLists.txt
Comment (tautschnig) [inline src/clobber/CMakeLists.txt]: Wishlist: Could those changes (and related ones) go in a separate commit, please?

File: src/clobber/clobber_parse_options.cpp
Comment (tautschnig) [inline src/clobber/clobber_parse_options.cpp]: Maybe a "Clobber: cleanup" commit?

File: src/goto-analyzer/CMakeLists.txt
Comment (tautschnig) [inline src/goto-analyzer/CMakeLists.txt]: To go in the cleanup commit.

File: src/clobber/Makefile
Comment (tautschnig) [inline src/clobber/Makefile]: Separate commit (see above).

File: src/ansi-c/ansi_c_language_info.h
Comment (tautschnig) [inline src/ansi-c/ansi_c_language_info.h]: Correct copy&paste, but we have meanwhile gone well beyond "ANSI" C.

File: src/goto-cc/goto_cc_languages.cpp
Comment (tautschnig) [inline src/goto-cc/goto_cc_languages.cpp]: The bottom part is for the cleanup commit.

File: src/goto-cc/goto_cc_languages.cpp
Comment (tautschnig) [inline src/goto-cc/goto_cc_languages.cpp]: The bottom part is for  the cleanup commit.

File: src/goto-diff/CMakeLists.txt
Comment (tautschnig) [inline src/goto-diff/CMakeLists.txt]: Cleanup commit.

File: src/goto-diff/goto_diff_languages.cpp
Comment (tautschnig) [inline src/goto-diff/goto_diff_languages.cpp]: Cleanup commit.

File: src/goto-diff/goto_diff_languages.cpp
Comment (tautschnig) [inline src/goto-diff/goto_diff_languages.cpp]: Cleanup commit.

File: src/jsil/jsil_language_info.h
Comment (tautschnig) [inline src/jsil/jsil_language_info.h]: Copy&paste error: "jsil" (should probably introduce an `ID_jsil`).

File: src/jsil/jsil_language_info.h
Comment (tautschnig) [inline src/jsil/jsil_language_info.h]: Copy&paste error, was "Javascript Intermediate Language"

File: src/langapi/language_info.h
Comment (tautschnig) [inline src/langapi/language_info.h]: Shouldn't these also be `= 0`? Those default implementations make no sense.

File: unit/analyses/dependence_graph.cpp
Comment (tautschnig) [inline unit/analyses/dependence_graph.cpp]: Nit pick: I'm not sure this is a value-adding change? We usually do grouping by directories.

File: unit/java_bytecode/java_object_factory/gen_nondet_string_init.cpp
Comment (tautschnig) [inline unit/java_bytecode/java_object_factory/gen_nondet_string_init.cpp]: As above: does this add value?

**Review by tautschnig — CHANGES_REQUESTED** [review]
I decided to ignore the don't-review-request... Just one genuine error spotted, otherwise nit-picks.

File: src/clobber/clobber_parse_options.cpp
Comment (peterschrammel) [reply src/clobber/clobber_parse_options.cpp]: Moved into a separate PR: https://github.com/diffblue/cbmc/pull/2159

**Comment by peterschrammel** [comment]
@tautschnig, this is ready for re-review.

**Comment by peterschrammel** [comment]
The TG bump needs to pass before merging this.

File: src/cbmc/cbmc_languages.cpp
Comment (tautschnig) [inline src/cbmc/cbmc_languages.cpp]: Would you be ok to add a commit to this PR that removes all occurrences of `HAVE_JSIL` from the codebase? It's not used consistently and thus doesn't really make sense. (Else it can happen in some later PR, once this is merged.)

**Review by tautschnig — APPROVED** [review]

File: src/langapi/language_info.h
Comment (tautschnig) [inline src/langapi/language_info.h]: Nit pick: you might use `= default;`

**Comment by peterschrammel** [comment]
@tautschnig, please check whether the last commit is what you intended.

**Comment by tautschnig** [comment]
> @tautschnig, please check whether the last commit is what you intended.

Yes, thank you very much!

**Comment by peterschrammel** [comment]
TG bump passing...

File: src/ansi-c/ansi_c_language_info.cpp
Comment (thk123) [inline src/ansi-c/ansi_c_language_info.cpp]: Should this be `Diffblue ltd`?

File: src/cpp/cpp_language_info.h
Comment (thk123) [inline src/cpp/cpp_language_info.h]: Either remove or turn into doxygen comment

File: src/java_bytecode/java_bytecode_language_info.cpp
Comment (thk123) [inline src/java_bytecode/java_bytecode_language_info.cpp]: And this should probably also be Diffblue ltd?

File: src/langapi/language_info.h
Comment (thk123) [inline src/langapi/language_info.h:48]: Why not `= 0`?

File: src/langapi/language_info.h
Comment (thk123) [inline src/langapi/language_info.h]: Please document these

File: src/langapi/mode.cpp
Comment (thk123) [inline src/langapi/mode.cpp:56]: Inner content spread over multiple lines so please wrap in braces

File: src/langapi/mode.cpp
Comment (thk123) [inline src/langapi/mode.cpp:70]: Inner content spread over multiple lines so please wrap in braces

Comment (thk123) [reply src/langapi/mode.cpp:70]: Also mild preference to use above method

File: src/langapi/mode.cpp
Comment (thk123) [inline src/langapi/mode.cpp]: Random fixes makes this long commit even harder to review - please avoid in future :slightly_smiling_face:

File: src/langapi/mode.cpp
Comment (thk123) [inline src/langapi/mode.cpp]: to avoid having two variables called `extension` (one abbreviated) could you rename either the original or the new one to something more descriptive? (e.g. `language_extension` vs `file_extension` perhaps?)

**Review by thk123 — COMMENTED** [review]
Just some minor changes, no need for re-review, will approve once I've checked TG bump

**Review by thk123 — APPROVED** [review]
TG pointer looks OK too

File: src/langapi/language_info.h
Comment (peterschrammel) [reply src/langapi/language_info.h:48]: Yes, that could be considered.

**Review by chrisr-diffblue — APPROVED** [review]
Looks like a reasonable cleanup to me.

File: src/cbmc/CMakeLists.txt
Comment (martin-cs) [inline src/cbmc/CMakeLists.txt:19]: The "remove conditional inclusion of JSIL" patch seems to be unrelated to the PR.  It may well be a good idea (Do we need conditional compilation of features?  If so, what is the policy?) but  at the moment it is not explained or justified.

File: src/ansi-c/ansi_c_language_info.cpp
Comment (martin-cs) [reply src/ansi-c/ansi_c_language_info.cpp]: If it is a new file written by a Diffblue employee, yes.

File: src/langapi/mode.cpp
Comment (martin-cs) [inline src/langapi/mode.cpp:77]: Unreachable?  Is this really an invariant?

**Review by martin-cs — APPROVED** [review]
Perhaps I am being unobservant but it seems like this could also tidy up some of the expr2c, expr2java, expr2internalstringinmadeuplanguage, expr2yetanotherformatbutwithdifferentprecedence mess.

File: src/cbmc/CMakeLists.txt
Comment (peterschrammel) [reply src/cbmc/CMakeLists.txt:19]: See thread above - requested by @tautschnig due to inconsistent use.

File: src/langapi/mode.cpp
Comment (peterschrammel) [reply src/langapi/mode.cpp:77]: You may call it that way, but I'd like to attach an error message to it ;)

**Review by mgudemann — APPROVED** [review]
LGTM

File: src/cbmc/CMakeLists.txt
Comment (tautschnig) [reply src/cbmc/CMakeLists.txt:19]: I agree that it would better fit a separate PR, but that PR would necessarily depend on this one because of overlapping changes, hence adding it in here seemed ok to me.

**Comment by kroening** [comment]
langapi is dead and deprecated!

**Review by smowton — APPROVED** [review]
Looks great, thanks for making the effort!

**Comment by smowton** [comment]
@kroening how would you prefer to do a language-specific thing, like "for the language relating to this symbol, get usual file extension" for example?

**Comment by peterschrammel** [comment]
@smowton, the suggestion is to move the language-specific dispatch from the global language registry to the driver program. I.e. everything that requires something language-specific will be parametrised by a function that performs the language-specific dispatch. The presumption is that this will only be necessary in very few cases (show_goto_trace, etc). The advantage is that we can get rid of a global variable.

**Comment by TGWDB** [comment]
This appears to be a nice to do, but is quite out of date. Closing since there seems to be a low likelihood of progress on this at the moment. Encourage the author to rebase and reopen if they would like this merged.

If you believe this closure is erroneous please reopen this PR.

**Pull request closed without merging**
